libiocage/lib/DevfsRules.py

```diff
@@ -0,0 +1,381 @@
+        self._ruleset_number_index = {}
+        self._ruleset_name_index = {}
+
+        # all line indexes that came from default rules (here called system-rules)
+        self._system_rule_lines = []
+
+
+        list.__init__(self)
+
+        # will automatically read from file - needs to be the last item
+        self.rules_file = rules_file
+
+    def append(self, ruleset, is_system_rule=False):
+        """
+        Add a DevfsRuleset to the list
+
+        The rulesets added become indexed, so that lookups and duplication
+        checks are easy and fast
+
+        Args:
+
+            ruleset (libiocage.lib.DevfsRules.DevfsRuleset|string):
+                The ruleset that gets added if it is not already in the list
+        """
+
+        next_line_index = len(self)
+
+        if ruleset is None or isinstance(ruleset, str):
+            list.append(self, ruleset)
+            if is_system_rule is True:
+                self._system_rule_lines.append(next_line_index)
+            return ruleset
+
+        if ruleset.name in self._ruleset_name_index.keys():
+            raise libiocage.lib.errors.DuplicateDevfsRuleset(
+                reason=f"Ruleset named '{ruleset.name}' already present",
+                devfs_rules_file=self.rules_file,
+                logger=self.logger
+            )
+
+        if ruleset.number in self._ruleset_number_index.keys():
+            raise libiocage.lib.errors.DuplicateDevfsRuleset(
+                reason=f"Ruleset number '{ruleset.number}' already present",
+                devfs_rules_file=self.rules_file,
+                logger=self.logger
+            )
+
+        # build indexes
+        self._ruleset_number_index[ruleset.number] = next_line_index
+        self._ruleset_name_index[ruleset.name] = next_line_index
+        if is_system_rule is True:
+            self._system_rule_lines.append(next_line_index)
+
+        list.append(self, ruleset)
+        return ruleset
+
+    def new_ruleset(self, ruleset):
+        """
+        Append a new ruleset
+
+        Similar to append(), but automatically assigns a new number
+
+        Args:
+
+            ruleset (libiocage.lib.DevfsRules.DevfsRuleset):
+                The new devfs ruleset that is going to be added
+
+        Returns:
+
+            int: The devfs ruleset number of the created ruleset
+        """
+
+        ruleset.number = self.next_number
+
+        if ruleset.name is None:
+            ruleset.name = f"iocage_auto_{ruleset.number}"
+
+        self.append(ruleset)
+        return ruleset.number
+
+    def find_by_name(self, rule_name):
+        return self._find_by_index(rule_name, self._ruleset_name_index)
+
+    def find_by_number(self, rule_number):
+        return self._find_by_index(rule_number, self._ruleset_number_index)
+
+    def _find_by_index(self, rule_name, index):
+        return self[index[rule_name]]
+
+    @property
+    def default_rules_file(self):
+        return "/etc/defaults/devfs.rules"
```

Comment:
Why are we using `/etc/defaults/devfs.rules` instead of `/etc/devfs.rules`?

Reply:
We are using both - "system" rules are loaded from `/etc/defaults/devfs.rules`.

That's important to be able to inherit rules from them. Have you noticed the real devfs_rule number provided by the user may be altered? In case the combination of iocage-required devfs rules and the user supplied rule (default `4`) has not been seen in this exact combination, a new rule is added permanently. When this happens, the devfs service is restarted.